### PR TITLE
Ruby enumerable methods

### PR DESCRIPTION
**Pull request by evabanegacom**
I created the branches and fixed the git-flow error that's why there are no many commits
i added the ruby file again and made the development branch the default one
This is some of ruby's enumerable methods, it was quite a headache some how i survived
i wrote each, select, any, none. etc.
these methods were written with rubocop guidelines and ruby itself.
I did some demonstration of the enumerable methods as well you can try them yourself

### CODE REVIEW

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  end
+
+  def my_all(*args)
+    my_each do |x|
+      return true unless block_given?
+
+      return false unless yield(x)
+    end
+    true
+  end
```
Comment (stratospherique) [inline enumerables.rb]: - [ ] Remane **my_all** to **my_all?** like in the original Enum module. :pray:

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      my_each { |x| return true if x }
+    else
+      my_each { |x| return true if x == yield(x) }
+    end
+    false
+  end
+
+  def my_none(*args)
+    !my_any?
+  end
```
Comment (stratospherique) [inline enumerables.rb]: The issues with **my_none?**, **my_any?** and **my_all?** are similar. the methods should be able to: 

* [x] Handle the cases where no block is given (Hint: block_given?)
* [ ] Return a suitable boolean result if no arguments were passed.
  ```ruby
	array = [false, true, false, nil, []]
	puts array.my_any? # should print true
	puts array.my_all? # should print false
	puts array.my_none? # should print false
  ```
* [ ] Accept primitives, classes (i.e Integer, String, etc...) and Regex Expressions as arguments.
  ```ruby
	words = %w[dog door rod blade]
	puts words.my_all?(/c/) == words.all?(/c/) # false
	puts words.my_any?(/c/) == words.all?(/c/) # false
	puts words.my_none?(/c/) == words.all?(/c/) # true
  ```

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    arguement = nil
+    if arguement
+      my_each { |x| counter += 1 if x == arguement }
+    elsif block_given?
+      my_each { |x| counter += 1 if yield x }
+    else
+      counter = length
+    end
+    counter
+  end
```
Comment (stratospherique) [inline enumerables.rb]: * [ ] should have working `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_count counts the number of items in enum that are equal to item if an argument is given
```ruby
array = [7,8,9,5,1,0,2,4,0,1]
array.my_count(0) == array.count(0) # false
```

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    element = element.to_sym if element.is_a?(String) && !Symbol && !Block
+    if element.is_a?(Symbol) && !symbol
+      block = element.to_proc
+      element = nil
+    end
+    symbol = symbol.to_sym if symbol.is_a?(String)
+    block = symbol.to_proc if symbol.is_a?(Symbol)
+    my_each { |x| element = element.nil? ? x : block.yield(element, x) }
+    element
+  end
```
Comment (stratospherique) [inline enumerables.rb]: * [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block
  ```ruby
    array = [4,7,8,89,9,6,5,12,4,2,5]
    operation = proc { |sum, n| sum + n  }
    p array.my_inject(&operation) == array.inject(&operation) # false
  ```

* [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  ```ruby
    array = [1,7,8,89,9,6,5,12,4,2,5]
    p array.my_inject(:+) == array.inject(:+) # return this error
    # NoMethodError:
        undefined method `yield' for nil:NilClass
      # ./enumerables.rb:95:in `block in my_inject'
      # ./enumerables.rb:8:in `block in my_each'
      # ./enumerables.rb:7:in `times'
      # ./enumerables.rb:7:in `my_each'
      # ./enumerables.rb:95:in `my_inject'
  ```

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great Start @evabanegacom :clap: :clap: However, some methods need a bit of refactoring in order for us to approve your project. Check :arrow_down: 

- **[REQUIRED]** The pull request is not showing any checks. :pray: fix it

- **[REQUIRED]** should have working `multiply_els` that uses `my_inject` as described on Odin ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/multiply_els.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- **[REQUIRED]** Check the code comments below, they should help you refactor your methods :gear: 

### That's it :relieved: When finished, kindly resubmit another code review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

**Comment by pelzolga123** [comment]
## Hi Precious!  
### You did a good job so far :+1: 
Still, there are changes required before this project can be approved:

*  #### My_all?
* [ ]  When a class is passed as an argument should return true if all of the collection is a member of such class. For example:
`[1, 'dog', []].my_all?(Integer)` - should return false

* [ ] When a pattern other than Regex or a Class is given should return true if all of the collection matches the pattern. For example:
`['dog','door','dish'].my_all?(/o/)` - should return false

* [ ] When a Regex is passed as an argument should return true if all of the collection matches the pattern. For example:
`[3, 2, 3].my_all?(3)` - should return false

*  #### My_any?
* [ ] Should return true if the block ever returns a value other then false or nil. For example:
`[1, 2, 3, 4].my_any?{|num| num <= 3}` - should return true

* [ ] When a class is passed as an argument should return true if at least one of the collection is a member of such class. For example:
`['dog', 'cat', 'bird'].my_any?(Integer)` - should return false

* [ ]  When a Regex is passed as an argument should return false if none of the collection matches the Regex. For example:
`['dog','door','dish'].my_any?(/z/)` - should return false

* [ ] When a pattern other than Regex or a Class is given should return false if none of the collection matches the pattern. For example:
`['dog', 'bird', 'fish'].my_any?('cat')` - should return false

*  #### My_none?
* [ ] Should return true if the block never returns true for all elements. For example:
`[1, 2, 3].my_none?{|num| num > 3}` - should return true

* [ ] When a class is passed as an argument should return true if none of the collection is a member of such class. For example:
`[1, 2 ,3].my_none?(String)` - should return true

* [ ]  When a Regex is passed as an argument should return true only if none of the collection matches the Regex. For example:
`['dog','door','dish'].my_none?(/z/)` - should return true

* [ ] When a pattern other than Regex or a Class is given should return true only if none of the collection matches the pattern. For example:
`['dog', 'bird', 'fish'].my_none?(5)` - should return true

### Please kindly fix the above issues and send another code review request.

If you have any questions or need help, you can message me on slack (@ Olga Pelts). *Happy coding*

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    #puts array.my_all? # should print false
+  #puts array.my_none? # should print false
+  #array = [7,8,9,5,1,0,2,4,0,1]
+p array.my_count(0) == array.count(0) # false
+words = %w[dog door rod blade]
+  # puts words.my_all?(/c/) == words.all?(/c/) # false
+  # puts words.my_any?(/c/) == words.all?(/c/) # false
+  # puts words.my_none?(/c/) == words.all?(/c/) # true
+  array = [1,7,8,89,9,6,5,12,4,2,5]
+  #p array.my_inject(:+) == array.inject(:+)
```
Comment (pelzolga123) [inline enumerables.rb]: No debugging instructions should be in committed code (as discussed [here](https://questions.microverse.org/t/debugging-instructions-code/346)) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/section/debugging_requirement.png) from Pathwright)

Comment (evabanegacom) [reply enumerables.rb]: okay are you on slack

File: enumerables.rb [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    array = nil
+    if array
+      my_each { |x| return false unless array == x }
+    elsif !block_given?
+      my_each { |x| return false unless x }
+    else
+      my_each { |x| return false unless yield(x) }
+    end
+    true
+  end
```
Comment (euqueme) [inline enumerables.rb]: * [ ] should have working `my_all?`  identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_all? 

- [ ] when a class is passed as an argument returns true only if all of the collection is a member of such class, otherwise, it should return false

```ruby     
[1, 2.5, 'a', 9].my_all?(Integer) == [1, 2.5, 'a', 9].all?(Integer) => false
```
- Yours is retrieving the wrong answer
image

- Real all?
image

- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex

```ruby     
%w[dog door rod blade].my_all?(/o/) == %w[dog door rod blade].all?(/o/) => false
```
- Yours is retrieving the wrong answer
image

- Real all?
image

- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

```ruby     
[3,4,7,11].my_all?(3) == [3,4,7,11].all?(3)  => false 
```
- Yours is retrieving the wrong answer
image

- Real all?
image

File: enumerables.rb [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    if args
+      my_each { |x| return true if x == args }
+    end
+    if !block_given?
+      my_each { |x| return true if x }
+    else
+      my_each { |x| return true if yield(x) }
+    end
+    false
+  end
```
Comment (euqueme) [inline enumerables.rb]: - [ ] should have working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

 #### my_any? 

- [ ] when a class is passed as an argument returns true if at least one of the collection is a member of such class

```ruby
[1, 2.5, 'a', 9].my_any?(Integer) == [1, 2.5, 'a', 9].any?(Integer) => true
```
- Yours is retrieving the wrong answer
my_any1

- Real any?
image

- [ ] when a Regex is passed as an argument returns false if none of the collection matches the Regex

```ruby
%w[dog door rod blade].my_any?(/o/) == %w[dog door rod blade].any?(/o/) => true
```
- Yours retrieves the wrong answer
image

- Real any?
image

- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

```ruby
%w[dog door rod blade].my_any?('cat') == %w[dog door rod blade].any?('cat')  => false
```
- Yours retrieves the wrong answer
image

- Real any?
image

File: enumerables.rb [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+  def my_none?(*args, &block)
+    !my_any?
+  end
```
Comment (euqueme) [inline enumerables.rb]: - [ ] should have working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_none?

- [ ] returns true if the block never returns true for all elements
```ruby
[0,5,4,9].my_none? {|num| num > 100} == [0,5,4,9].none? {|num| num > 100} => true
```
- Yours is retrieving thew wrong answer
image

- Real none?
image

- [ ] when a class is passed as an argument returns true if none of the collection is a member of such class
```ruby
[1, 2.5, 9].my_none?(String) == [1, 2.5, 9].none?(String) => true
```
- Yours is retrieving the wrong answer
image

- Real none?
image

- [ ] when a Regex is passed as an argument returns true only if none of the collection matches the Regex
```ruby
%w[dog door rod blade].my_none?(/z/) == %w[dog door rod blade].none?(/z/) => true
```
- Yours is retrieving the wrong answer
image

- Real none?
image

- [ ] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
```ruby
[4,7,11].my_none?(3) == [4,7,11].none?(3) => true
```
- Yours is retrieving the wrong answer
image

- Real none?
image

File: enumerables.rb [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    element = element.to_sym if element.is_a?(String) && !Symbol && !Block
+    if element.is_a?(Symbol) && !symbol
+      block = element.to_proc
+      element = nil
+    end
+    symbol = symbol.to_sym if symbol.is_a?(String)
+    block = symbol.to_proc if symbol.is_a?(Symbol)
+    my_each { |x| element = element.nil? ? x : block.yield(element, x) }
+    element
+  end
```
Comment (euqueme) [inline enumerables.rb]: - [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

####  my_inject

- [ ] combines all elements of enum by applying a binary operation, specified by a block
```ruby
(1..5).my_inject(4) { |prod, n| prod * n } == (1..5).inject(4) { |prod, n| prod * n } => 480
```
- Yours is retrieving thew wrong answer
image

- Real inject
image

- [ ] When a symbol is provided as an argument, my_inject should combine the elements of the enum using that symbol as a named method.
```ruby
[2, 5, 6].my_inject(:+) == [2, 5, 6].inject(:+) => 13
```
- Yours is retrieving a NoMethodError
image

- Real inject
image

File: enumerables.rb [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    arrays = Array
+    array = arrays(item)
+    array.my_inject { |item1, item2| item1 * item2 }
+  end
+
+ p  [1, 2, 3, 4].all?{|num| num <= 3} #should return true
+
+ p [1, 'dog', []].all?(Integer) - #should return false
+
+ p ['dog','door','dish'].all?(/o/) - #should return false
```
Comment (euqueme) [inline enumerables.rb]: - [ ] delete the '-' at the end of these lines it's making your file retrieve errors

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @evabanegacom :handshake: 

### Thank you for doing the changes requested by the previous reviewer, :medal_sports: almost all of your enumerable methods are working fine and retrieving no execution errors but a few are retrieving the wrong answer and only one of them an execution error, please address to the comments I've left in your code

**Remember:**  to follow this [guide](https://ruby-doc.org/core-2.6.5/Enumerable.html), all the examples for each of the Enumerable methods should be able to work using your own Enumerable methods.

### That's it! :raised_hands: 

### Please :pray: do the changes :man_technologist: and submit back for a code review when you are done :memo: 

Happy Coding :rocket: 

Any questions? is better to ask me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :vertical_traffic_light:

**Comment by evabanegacom** [comment]
i can analyze these methods to an extent even in plain English if necessary. I'm not sure where i'm not getting it correct. But i'll keep trying thanks for the reviews. If i'm wrong do not hesitate to tell me it would help me learn more

File: enumerables.rb [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+  def my_all?(array = nil)
+    if array
+      my_each { |x| return false unless array == x }
+    elsif !block_given?
+      my_each { |x| return false unless x }
+    else
+      my_each { |x| return false unless yield(x) }
+    end
+    true
+  end
```
Comment (euqueme) [inline enumerables.rb]: * [ ] should have working `my_all?`  identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_all? 

- [ ] when a class is passed as an argument returns true only if all of the collection is a member of such class, otherwise, it should return false

```ruby     
[1, 2i, 3.14].my_all?(Numeric) == [1, 2i, 3.14].all?(Numeric) => true
```
- Yours is retrieving the wrong answer
image

- Real all?
image

- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex

```ruby     
%w[dog door rod blade].my_all?(/d/) == %w[dog door rod blade].all?(/d/) => true
```
- Yours is retrieving the wrong answer
image

- Real all?
image

File: enumerables.rb [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    if args
+      my_each { |x| return false unless args == x }
+    end
+    if !block_given?
+      my_each { |x| return true if x }
+    else
+      my_each { |x| return true if yield(x) }
+    end
+    false
+  end
```
Comment (euqueme) [inline enumerables.rb]: - [ ] should have working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

 #### my_any? 

- [ ] when a class is passed as an argument returns true if at least one of the collection is a member of such class

```ruby
[1, 2.5, 'a', 9].my_any?(Integer) == [1, 2.5, 'a', 9].any?(Integer) => true
```
- Yours is retrieving the wrong answer
my_any1

- Real any?
image

- [ ] when a Regex is passed as an argument returns false if none of the collection matches the Regex

```ruby
%w[dog door rod blade].my_any?(/o/) == %w[dog door rod blade].any?(/o/) => true
```
- Yours retrieves the wrong answer
image

- Real any?
image

- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

```ruby
%w[dog door rod blade cat].my_any?('cat') == %w[dog door rod blade cat].any?('cat')  => true
```
- Yours retrieves the wrong answer
image

- Real any?
image

File: enumerables.rb [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+
+  def my_none?(args = nil, &block)
+    !my_any?(args, &block)
+  end
```
Comment (euqueme) [inline enumerables.rb:59]: - [ ] should have working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- I see you are getting my_none? from my_any? so once you solve the issues with your my_any? your my_none? will be solved as well

File: enumerables.rb [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+    element = element.to_sym if element.is_a?(String) && !symbol && !block
+    if element.is_a?(Symbol) && !symbol
+      block = element.to_proc
+      element = nil
+    end
+    symbol = symbol.to_sym if symbol.is_a?(String)
+    block = symbol.to_proc if symbol.is_a?(Symbol)
+    my_each { |x| element = element.nil? ? x : block.yield(element, x) }
+    element
+  end
```
Comment (euqueme) [inline enumerables.rb]: - [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

####  my_inject

- [ ] combines all elements of enum by applying a binary operation, specified by a block
```ruby
(1..5).my_inject(4) { |prod, n| prod * n } == (1..5).inject(4) { |prod, n| prod * n } => 480
```
- Yours is retrieving a NameError
image

- Real inject
image

- [ ] When a symbol is provided as an argument, my_inject should combine the elements of the enum using that symbol as a named method.
```ruby
[2, 5, 6].my_inject(:+) == [2, 5, 6].inject(:+) => 13
```
- Yours is retrieving a NoMethodError
image

- Real inject
image

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hi @evabanegacom :wave: 

### I see you've solved some of the changes I requested earlier, so thank you for the great effort you've put into this project :muscle: :medal_sports: we still need to work on some changes :construction: :construction_worker: so l have left comments in your code below :point_down: 

## Project Specifications

* [ ] should have working `multiply_els` that uses `my_inject` as described on Odin ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/multiply_els.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- I'm sorry I completely missed this on my last review :bowing_woman: :disappointed:  but the first reviewer mention it so please work on the multiply_els method

## General
* [ ] Linters config:  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/linter_config.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/54883773/))

- Again I'm sorry I didn't see this :bowing_woman: :disappointed:  but the first reviewer noticed it, I see you have your linter and stickler files so probably you only need to activate the repo on [stickler](https://stickler-ci.com/), nevertheless,  do not :no_good:  submit for a code review if you don't see your last commit with a green checkmark :heavy_check_mark: 

### That's it! :raised_hands: 

### Please :pray: do the changes :technologist: and submit back for a code review when you are done :memo: 

Happy Coding! :rocket: 

Any question? I'd be happy to help on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :vertical_traffic_light:

**Comment by evabanegacom** [comment]
This thread is really getting long lol

**Review by whiz25 — APPROVED** [review]
## Approved :clap: 
Hey, @evabanegacom you have come a long way with this project. There are still a few failing edge cases but I am approving this PR since it has surpassed the max review cycle. Merge and proceed to the next projects.

### Martin Afani

**Comment by evabanegacom** [comment]
Thanks, but, i really didn't mind working on it, nonetheless i'll look to correct the flaws
